Support for detailed error messages on presto-client side (+CLI support)

fiedukow opened the pull request:
This change tries to make message of some known problems more informative for presto users.
The message does not cover all possibilities, however it covers our best known scenario for that exception, so it would be nice to have that tips at this point.
Also technical details are hidden and will require `--debug` to be shown. Note that stacktrace was visible only with `--debug` even before this change.

kokosing commented on presto-cli/src/main/java/com/facebook/presto/cli/Console.java:
```diff
@@ -322,11 +324,41 @@ private static void process(QueryRunner queryRunner, String sql, OutputFormat ou
+        }
+    }
+
+    private static String getErrorMessage(RuntimeException e, ClientSession session)
```
`getErrorMessage` suggest that you will retrieve kind of an error message from exception, while you create something more than error message

electrum replied:
Make this `Throwable throwable`

fiedukow replied:
Does `createErrorMessage` sounds better?

Changed RuntimeException to throwable.

kokosing commented on presto-cli/src/main/java/com/facebook/presto/cli/Console.java:
```diff
@@ -322,11 +324,41 @@ private static void process(QueryRunner queryRunner, String sql, OutputFormat ou
+    {
+        StringWriter sw = new StringWriter();
+        e.printStackTrace(new PrintWriter(sw));
+        String stackTrace = sw.toString();
```
I would wrapt this as static method utility method `getStackTrace(e)`, moreove this variable is used in condition block so I would create this variable there

fiedukow replied:
done

kokosing commented on presto-cli/src/main/java/com/facebook/presto/cli/Console.java:
```diff
@@ -322,11 +324,41 @@ private static void process(QueryRunner queryRunner, String sql, OutputFormat ou
+        e.printStackTrace(new PrintWriter(sw));
+        String stackTrace = sw.toString();
+
+        StringBuilder builder = new StringBuilder();
```
I am not sure if using builder here is needed. It is not performance crucial part of the code. I would use plain `+` operator

fiedukow replied:
Whats the downside of builder?

kokosing replied:
verbose code, anyway it is a minor comment. up to you

kokosing commented on presto-cli/src/main/java/com/facebook/presto/cli/Console.java:
```diff
@@ -322,11 +324,41 @@ private static void process(QueryRunner queryRunner, String sql, OutputFormat ou
+        String stackTrace = sw.toString();
+
+        StringBuilder builder = new StringBuilder();
+        if (e.getCause() instanceof java.io.EOFException) {
```
use import

electrum replied:
This should also include `ConnectException`

electrum replied:
Using `Throwables.getCausalChain` is safer than checking only the first cause

fiedukow replied:
done/done/done

kokosing commented:
lgtm % minor comments

electrum commented on presto-cli/src/main/java/com/facebook/presto/cli/Console.java:
```diff
@@ -322,11 +324,41 @@ private static void process(QueryRunner queryRunner, String sql, OutputFormat ou
+        }
+
+        if (session.isDebug()) {
+            builder.append("\n=========   TECHNICAL DETAILS   =========\n");
```
I'd rather not add this since this is debug mode, where you expect "technical details" to be printed (otherwise you wouldn't have specified `--debug`).

fiedukow replied:
The output just looks nicer with that and it's easier to find what you want if you are human reading this. For any possible auto-analysis it shouldn't really meter.

electrum commented on presto-cli/src/main/java/com/facebook/presto/cli/Console.java:
```diff
@@ -322,11 +324,41 @@ private static void process(QueryRunner queryRunner, String sql, OutputFormat ou
+
+    private static String getErrorMessage(RuntimeException e, ClientSession session)
+    {
+        StringWriter sw = new StringWriter();
```
Name this `writer`

fiedukow replied:
done

electrum commented on presto-cli/src/main/java/com/facebook/presto/cli/Console.java:
```diff
@@ -322,11 +324,41 @@ private static void process(QueryRunner queryRunner, String sql, OutputFormat ou
+
+        StringBuilder builder = new StringBuilder();
+        if (e.getCause() instanceof java.io.EOFException) {
+            builder.append("There was a problem with a response from Presto Coordinator." + "\n");
```
These string literals can be joined.

The phrasing here is a bit off. I'll improve it before merging.

fiedukow replied:
Thanks :)

electrum commented:
Can we make this work for 404 errors as well, which happens when someone points the CLI at a non-Presto HTTP service? (this has come up a few times)

fiedukow commented:
As for the 404, I would add it as another task to do for TD, as we will probably want to improve even more error messages in CLI. Is that acceptable?

Comments applied.

fiedukow commented:
Thanks for the review.

Rebased to current master - back to you @electrum

fiedukow commented:
ping

fiedukow commented:
I rewritten whole thing.
Support for `404 error` (as requested) as well as for `server still initializing` is added.
I've introduced `--quiet` flag to lower down verbosity when required.
Here is how error messages do look right now for CLI with different flags:
https://gist.github.com/fiedukow/eb71b76411e739128032374b7723ecde

@electrum Is that satisfactory to you now?
@KBP-TDC Can you review error messages?
@kokosing Any new comments with new version of code?

kokosing commented on presto-spi/src/main/java/com/facebook/presto/spi/PrestoException.java:
```diff
@@ -57,4 +68,47 @@ public String getMessage()
+        return writer.toString();
+    }
+
+    public static class SerializedPrestoException
```
it is not Exception as name suggests

kokosing commented on presto-spi/src/main/java/com/facebook/presto/spi/PrestoException.java:
```diff
@@ -40,6 +46,11 @@ public PrestoException(ErrorCodeSupplier errorCodeSupplier, String message, Thro
         this.errorCode = errorCodeSupplier.toErrorCode();
     }
 
+    public SerializedPrestoException getSerialized()
```
it is not a getter so I would name it `toSerialized()` or even without shortcut `toSerializedPrestoException()`

kokosing commented on presto-client/src/main/java/com/facebook/presto/client/StatementClient.java:
```diff
@@ -338,12 +341,19 @@ private void processResponse(JsonResponse<QueryResults> response)
         }
-        return new RuntimeException(format("Error %s at %s returned %s: %s", task, request.getUri(), response.getStatusCode(), response.getStatusMessage()));
+        catch (IOException e) {
+            // The response does not fit into PrestoException, continue with adding just request.
```
maybe add debug logging here, just in case

fiedukow replied:
discussed offline to leave it as is

kokosing commented on presto-cli/src/main/java/com/facebook/presto/cli/ErrorMessages.java:
```diff
@@ -16,17 +16,72 @@
     private static final String RESPONSE_INTRO = "[ Plain text HTTP response ]\n";
     private static final String TECHNICAL_DETAILS_END = "========= TECHNICAL DETAILS END =========\n\n";
 
+    private enum Tips {
```
just `Tip`?

kokosing commented on presto-cli/src/main/java/com/facebook/presto/cli/ErrorMessages.java:
```diff
@@ -88,7 +90,12 @@ public Builder addTip(Tips tip, Object ...toFormat)
             public String build()
             {
-                return builder.toString();
+                if (!session.isQuiet()) {
```
I would prefer to use `quiet` in the same manner as `debug` is used. So in case when `quiet` is enabled then I would not add any `Tip` to builder. 

What happens when `--debug --quiet` is used?
Make sure you don't display TIP_INTRO when no `Tip` is added.

fiedukow replied:
`--debug` + `--quiet` will enable technical details but disable tips. I didn't want to spend to much time thinking about other possible solutions for that as that seemed to be good enough.

:+1: TIP_INTRO
:+1: use `quiet` as `debug`

kokosing commented:
looks good % minor comments

fiedukow commented:
@kokosing Comments applied.
@electrum Can you share your thoughts on that?

electrum commented on presto-cli/src/main/java/com/facebook/presto/cli/ErrorMessages.java:
```diff
@@ -0,0 +1,45 @@
+    private static String getStackTraceString(Throwable throwable)
+    {
+        StringBuilder builder = new StringBuilder();
+        builder.append(Arrays.stream(throwable.getStackTrace()).map(x -> x.toString()).reduce("", (x, y) -> x + y + "\n"));
```
This is really hard to read

fiedukow replied:
fixed

fiedukow commented:
Rebase to master, as the previous base was not passing all product tests.

fiedukow commented:
Rebased to master. Conflicts solved.

fiedukow commented:
Rebased to master. Conflicts solved.

electrum commented:
Can you pull the serialization change into a separate pull request? This is a protocol change, so it will be best to review and merge it separately. I have some specific thoughts/comments on it (we do something similar in our internal gateway / query router).

fiedukow commented:
This commit is kinda pointless without the context of this whole PR, but here is it:
https://github.com/prestodb/presto/pull/6199
Extracted as requested.

I left the commit here as well though as it's strictly related to rest of the PR content. I will drop it from PR when the serialization will land on master.

fiedukow commented:
I rebased whole PR to master with @dain code that causes `Server Still Initializing` to be returned as QueryError.
I've made major changes in the code adjusting it to usage of QueryError instead of PrestoException.

I think we can carry on with review as it looks like #6658 will be integrated soon.

dain commented:
@fiedukow is this still needed now that #6658 has been merged?

fiedukow commented:
Absolutely, this PR is not only about sending proper message to client when session was not set, but also to provide nicely formatted messages for users to increase ease of use.
Those messages are added for the server starting up case, as well as many others.
Also there is a place to add more in the future.

fiedukow commented:
Rebased to master, dropped @dain commit as it's already integrated to master.

dain commented:
@fiedukow most of this change falls into the category of "product decisions" (e.g, how do we want to format this? do we need all of these fields? are the messages good?) and will need @electrum to review.  

@electrum if you only want to do the "product design" parts, I can review the code... just assign it back to me

fiedukow commented on presto-cli/src/main/java/com/facebook/presto/cli/ErrorMessages.java:
```diff
@@ -0,0 +1,45 @@
+            builder.append("caused by:\n");
+            builder.append(getStackTraceString(throwable.getCause()));
+        }
+        return builder.toString();
```
This was just a leftover, in new version I don't implement printing stack by myself.
I change this commit to get rid of this leftover, so it should not be visible in any commit now.

fiedukow commented:
@dain 
Of course this will require @electrum review, also @electrum kinda did that already at the beginning (15 June 2016). I added you mostly to confirm that I used your patch as intended.

@electrum
I hope you will be able to finish that review now.

fiedukow commented:
Kind ping @electrum

fiedukow commented:
Kind ping again :)

fiedukow commented:
Kind ping. Can I do something to push that forward?

findepi commented:
Current `master` looks terribly:
```
$ presto-cli-*-executable.jar --server ec2-xxx-xxx-xxx-xxx.compute.amazonaws.com:<wrong port>
presto> select 1;
<... 5 seconds elapsed ...>
Error running command: java.util.concurrent.TimeoutException: Total timeout 5000 ms elapsed
```

With all the possible problems there actually can be, `TimeoutException: Total timeout 5000 ms elapsed` error did in no  way help me understand I have a wrong port and no connection could be established. 

@electrum, when is this PR going to be merged?

fiedukow commented:
Rebased to master and resolved conflicts

fiedukow commented:
ping?

fiedukow commented:
@electrum @dain @martint This is really simple code and concept and it is going to be another year old PR in less then two weeks. Could you please decide if you want that and if so just merge the code or tell me what should I change. I'm not buying any more cakes for my PRs birthdays anyway ;)

martint commented on presto-cli/src/main/java/com/facebook/presto/cli/ErrorMessages.java:
```diff
@@ -0,0 +1,42 @@
+
+public class ErrorMessages
+{
+    private ErrorMessages()
```
We generally format this as:

```java
private ErrorMessages() {}
```

or 
```java
private ErrorMessages()
{
}
```

martint commented on presto-cli/src/main/java/com/facebook/presto/cli/ErrorMessages.java:
```diff
@@ -66,13 +66,6 @@ private static String runtimeExceptionErrorMessage(Throwable throwable, ClientSe
         return builder.toString();
     }
 
-    private static String getStackTraceString(Throwable throwable)
```
Unrelated change? Why did this move? (It make it harder to review, since it's not clear whether there are any actual changes)

fiedukow replied:
It was unrelated change. It's gone now, as this method doesn't exists anymore.

martint commented on presto-cli/src/main/java/com/facebook/presto/cli/ErrorMessages.java:
```diff
@@ -14,22 +14,55 @@
     {}
 
-    public static String createErrorMessage(Throwable throwable, ClientSession session)
+    public static String createQueryErrorMessage(StatementClient client)
```
It seems strange for this class to have a dependency on StatementClient. I would change it to just depend on the error and let the caller extract it.

fiedukow replied:
It's not only about error. This class is very specific and depends on many other classes anyway. It is meant to recognise multiple specific exception classes that only make sense in the context of Presto.

This particular dependency allow to provide better error messages by extracting informations like query, debug mode, and more. Generalising that class is a cool project for an external library, I didn't like the idea of over engineering simple things we want to do in this class though.

If you prefer generalised way of doing that that will (for example) allow to map particular parameters from outside of class and make the process of error recognition dynamic, please let me know - I can do that with pleasure - I just didn't thought it's worth it :)

I can also to something in between and only get rid of StatementClient dependency, but my opinion is still to leave it as is.

Please let me know what you decide.

martint commented on presto-cli/src/main/java/com/facebook/presto/cli/ErrorMessages.java:
```diff
@@ -14,22 +14,55 @@
+        return standardQueryErrorMessage(client);
+    }
+
+    public static String createExceptionMessage(Throwable throwable, ClientSession session)
```
Why did the name of this method change?

fiedukow replied:
I wanted to make the difference in what does this method do in comparison to createQueryErrorMessage is clear.

martint commented on presto-cli/src/main/java/com/facebook/presto/cli/ErrorMessages.java:
```diff
@@ -14,22 +14,55 @@
+        return runtimeExceptionErrorMessage(throwable, session);
+    }
+
+    private static String runtimeExceptionErrorMessage(Throwable throwable, ClientSession session)
```
This seems to be used in a single place in the method above. Inline it.

martint commented on presto-cli/src/main/java/com/facebook/presto/cli/ErrorMessages.java:
```diff
@@ -39,4 +72,80 @@ private static String getStackTraceString(Throwable throwable)
+
+    private static void technicalDetailsRuntimeExceptionErrorMessage(StringBuilder builder, Throwable throwable, ClientSession session)
+    {
+        builder.append(TECHNICAL_DETAILS_HEADER);
```
These constants are used only here, so I don't see the benefit of having them extracted. It makes it harder to tell exactly what's going to be printed, too. Inline them.

martint commented on presto-cli/src/main/java/com/facebook/presto/cli/ErrorMessages.java:
```diff
@@ -0,0 +1,42 @@
+        return builder.toString();
+    }
+
+    private static String getStackTraceString(Throwable throwable)
```
Guava's Throwables.getStackTraceAsString does this already.

martint commented on presto-cli/src/main/java/com/facebook/presto/cli/ErrorMessages.java:
```diff
@@ -39,4 +72,80 @@ private static String getStackTraceString(Throwable throwable)
+
+        builder.append(String.format("Query %s failed: %s%n", client.finalResults().getId(), error.getMessage()));
+        if (client.isDebug() && (error.getFailureInfo() != null)) {
+            StringWriter errors = new StringWriter();
```
Isn't this just the same as Guava's Throwables.getStackTraceAsString?

martint commented on presto-cli/src/main/java/com/facebook/presto/cli/ErrorMessages.java:
```diff
@@ -39,4 +72,80 @@ private static String getStackTraceString(Throwable throwable)
+            bad = " <EOF>";
+        }
+
+        if (REAL_TERMINAL) {
```
This class shouldn't depend on ConsolePrinter, and shouldn't be trying to determine whether it's being used to output to a terminal in this way (e.g., it's possible that the console is outputting to the terminal, but this class is being invoked to render an error that will go to a log file).

A better way would be for this method to take an argument indicating whether it should output ANSI escape codes or not.

martint commented on presto-cli/src/main/java/com/facebook/presto/cli/ErrorMessages.java:
```diff
@@ -23,22 +23,75 @@
 
+    private enum Tip {
+        VERIFY_PRESTO_RUNNING("Verify that Presto is running on %s."),
+        DEFINE_SERVER_AS_CLI_PARAM("Use '--server' argument when starting Presto CLI to define server host and port."),
```
One concern about these is that, at FB, we wrap the CLI with a launcher that sets some parameter (server, etc. after looking up the corresponding cluster given a symbolic name). These hints are not relevant to our users, and may actually be confusing. 

E.g.,
- they have no way to verify presto is running on xxx. That's something managed by the Presto team. The message we would show in this case would be "Check XXX internal group for announcements about maintenance", or "check XXX status page for availability information".
- they can't provide the --server argument. That's handled by the wrapper. Setting it would cause the wrapper to fail to run.
- they can't check network conditions between client and server. That's infrastructure they have no control over or knowledge of how it works.
- it's unlikely they'll understand the output of --debug or be able to make any sense of it.

fiedukow replied:
I understand that concern.
Isn't `--quiet` mode (introduced later in this PR) an answer for you? In quiet mode only the main message is displayed and all tips are hidden. Wouldn't adding that flag to CLI while starting it through your wrapper be a solution to that?

fiedukow commented:
I applied/responded to all your comments.
I've added changes as fixups (except of resolving conflicts ofc) so you can see what I've done.
Back to you @martint

fiedukow commented:
This requires another turn of solving conflicts, but I don't see a point in doing that if we are not going to merge it.
Can you review further commits or at least accept the feature as it is so someone from TD commiters can continue with code review?

findepi commented:
Superseded by #9326.
